metagraph/src/common/file_merger.hpp

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include <algorithm>
+#include <filesystem>
+#include <fstream>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include "common/logger.hpp"
+
+namespace mtg {
+namespace common {
+
+/**
+ * Heap implemented as a sorted vector.
+ *
+ * The heap uses a vector as the underlying data structure, thus making it efficient
+ * only for a small (<1000) number of elements.
+ * @tparam T the actual heap element type; this is the type tested for equality
+ */
+// Note: profiling shows that using a sorted vector instead of a std::priority_queue is
+// faster up to ~1000 elements.  Using an unsorted vector (faster insert,
+// slower pop()) is ~40% slower. Preventing duplicates in the heap so that we don't
+// need to test for dupes at pop time is ~60% slower.
+// Note: Profiling indicates that merging elements within the heap is ~30% slower than
+// inserting duplicates and merging them when popping from the heap, as we do now
+template <typename T, class Compare = std::greater<T>>
+class MergeHeap {
+    /** The heap stores pairs <Element, SourceIndex> */
+    using value_type = std::pair<T, uint32_t>;
+
+  public:
+    inline void emplace(T el, uint32_t idx) {
+        els.emplace_back(el, idx);
+        std::push_heap(els.begin(), els.end(),
+                       [&](const value_type &a, const value_type &b) {
+                           return compare_(a.first, b.first);
+                       });
+    }
+
+    inline const value_type& top() const { return els.front(); }
+
+    inline value_type pop() {
+        std::pop_heap(els.begin(), els.end(),
+                      [&](const value_type &a, const value_type &b) {
+                          return compare_(a.first, b.first);
+                      });
+        value_type result = els.back();
+        els.pop_back();
+        return result;
+    }
+
+    inline bool empty() const { return els.empty(); }
+
+  private:
+    // elements stored as a min-heap
+    std::vector<value_type> els;
+    Compare compare_ = Compare();
+};
```

Comment:
Why do we need yet another merger? Why not use the ones from Elias Fano mergers?

Reply:
I wanted a simpler merger that doesn't compress/decompress data, as in this case disk space is not an issue. But we can totally use the EF encoders, and the corresponding merger, too.